### PR TITLE
Fix/speciestree

### PR DESCRIPTION
**Pull request by CristiGuijarro**
The species tree worked perfectly for the large default species_sets such as vertebrates but was not set up to work for smaller species_sets such as 10 rice species. There were complications with the outgroup and species_set parameters conflicting with the collection parameters, now only the outgroup species name is needed, the genome_db_id will be collected through the genome_db adaptor. The root is also now decided based on species_set rather than forced to be the tree of life ancestral root. The un-rooting step worked well for the large trees, but was unable to unroot the smaller trees without a forced midpoint rooting first.

### CODE REVIEW

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/GraftSubtrees.pm [hunk 1]
```diff
@@ -59,8 +59,8 @@ sub run {
 
 	my %trees = %{ $self->param('trees')}; # assuming { group_id => { tree => newick_string, outgroup => outgroup_id } }
 
-	# my $root_tree = $trees{root}->{tree};
-	my $root_tree = $trees{1}->{tree};
+	my $root_tree = $trees{$self->param_required('root_id')}->{tree};
+	# my $root_tree = $trees{1}->{tree};
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/GraftSubtrees.pm]: Remove this commented line

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/GroupSpecies.pm [hunk 2]
```diff
@@ -67,24 +67,29 @@ sub fetch_input {
 	my $species_set_id = $self->param('species_set_id');
 
 	die "Need either collection or species_set_id to be defined!" unless ( $collection || $species_set_id );
+  #die "collection and species_set_id are mutually exclusive, please pick only one" if ( $collection && $species_set_id );
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/GroupSpecies.pm]: Remove this commented line

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/GroupSpecies.pm [hunk 3]
```diff
@@ -122,6 +127,11 @@ sub write_output {
 		$self->dataflow_output_id( $group_set, 1 );
 		$c++;
 	}
+  
+  $self->dataflow_output_id( {
+    'param_name' =>'outgroup_id',
+    'param_value' => $self->param('outgroup_genome_db_id')
+  }, 2 );
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/GroupSpecies.pm]:
```suggestion
  }, 2 ); # to pipeline_wide_parameters
```

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/GroupSpecies.pm [hunk 4]
```diff
@@ -67,24 +67,29 @@ sub fetch_input {
...
 
 	my $dba = $self->compara_dba;
 	my $gdb_adaptor = $dba->get_GenomeDBAdaptor;
 	my $ss_adaptor = $dba->get_SpeciesSetAdaptor;
 	
 	my $species_set;
-	if ( $collection ) {
-		$species_set = $ss_adaptor->fetch_collection_by_name($collection);
-	} elsif ( $species_set_id ) {
+  if ( $species_set_id ) {
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/GroupSpecies.pm]: Fix indentation here

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/GroupSpecies.pm [hunk 5]
```diff
@@ -67,24 +67,29 @@ sub fetch_input {
...
 
 	my @gdb_id_list = map { $_->dbID } @genome_dbs;
-	push( @gdb_id_list, @{ $self->param('outgroup_gdbs') } ) if $self->param('outgroup_gdbs');
 	@gdb_id_list = sort {$a <=> $b} @gdb_id_list; 
 	$self->param( 'dataflow_groups', [{ genome_db_ids => \@gdb_id_list, collection => $self->param('collection')}] );
+  
+  #Find outgroup_id
+  my $outgroup_name = $self->param('outgroup');
+  my $outgroup_genome_db = $gdb_adaptor->fetch_by_name_assembly($outgroup_name);
+  $self->param('outgroup_genome_db_id', $outgroup_genome_db->dbID);
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/GroupSpecies.pm]: Fix indentation here

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm [hunk 6]
```diff
@@ -121,11 +130,13 @@ sub fetch_input {
 			$submatrix = $submatrix->collapse_group_in_matrix( $prev_group_gdbs, "mrg_$prev_group" );
 		}
 
+    next if Bio::EnsEMBL::Compara::Utils::DistanceMatrix->empty_submatrix($submatrix);
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm]: Fix indentation

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm [hunk 7]
```diff
@@ -96,16 +97,24 @@ sub fetch_input {
 
 	my (@prev_group_ids, %all_groups, @matrices_dataflow, $this_outgroup);
 	foreach my $group_taxon_id ( @tax_groups_ids ) {
+    # $group_taxon_id=4527;
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm]: Remove commented line

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm [hunk 8]
```diff
@@ -96,16 +97,24 @@ sub fetch_input {
...
 		my $current_group_taxon = $ncbi_adaptor->fetch_node_by_taxon_id($group_taxon_id);
 		print " -- Grouping " . $current_group_taxon->name . "...\n" if $self->debug;
 
 		# extract initial submatrix for the group
 		my @group_gdbs = @{$gdb_adaptor->fetch_all_current_by_ancestral_taxon_id($group_taxon_id)};
+    #print Dumper @group_gdbs;
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm]: Remove commented line

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm [hunk 9]
```diff
@@ -144,6 +155,7 @@ sub write_output {
 	#     { group_key => $key2, distance_matrix => $matrix2 }, # no outgroup
 	# ]
 	$self->dataflow_output_id( $self->param('matrices_dataflow'), 2 );
+  $self->dataflow_output_id( { root_id => $self->param('root_id') }, 1 );
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm]: Fix indentation

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm [hunk 10]
```diff
@@ -154,9 +166,10 @@ sub _add_outgroup {
 
 	my ( $closest_gdb_id, $min_distance) = (undef, 100);
 	foreach my $full_key ( $full_matrix->members ) {
+    print "FULL_KEY: $full_key\n";
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm]: Remove debugging print

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm [hunk 11]
```diff
@@ -154,9 +166,10 @@ sub _add_outgroup {
...
 		next if ( defined $self->param('blacklisted_genome_db_ids') && grep { $full_key eq $_ } @{$self->param('blacklisted_genome_db_ids')} ); # skip any that are on the naughty list
 		next if (grep { $_ eq $full_key } @sub_gdb_ids); # skip any that exist in the submatrix
-
+    next unless defined $full_matrix->distance($rep_gdb_id, $full_key);
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm]: Not sure about this line... Is it needed? What happens without it? Should it `die` if there's no distance between some members of the submatrix instead?

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm [hunk 12]
```diff
@@ -235,8 +257,12 @@ sub _taxonomic_groups {
 	my @sorted_taxon_ids = map {$_->dbID} @sorted_taxa;
 
 	# finally, add the root
-	my $root_taxon = $ncbi_adaptor->fetch_node_by_name('root');
+  my $largest_group = $sorted_taxa[-1];
+  my $root_taxon = $ncbi_adaptor->fetch_parent_for_node($largest_group);
+  
+	# my $root_taxon = $ncbi_adaptor->fetch_node_by_name('root');
 	push(@sorted_taxon_ids, $root_taxon->dbID); # usually 1, but fetch from db just incase
+  $self->param('root_id', $root_taxon->dbID);
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm]: Indentation

File: modules/Bio/EnsEMBL/Compara/Utils/DistanceMatrix.pm [hunk 13]
```diff
@@ -223,11 +223,15 @@ sub add_taxonomic_distance {
...
 sub prune_gdbs_from_matrix {
 	my ( $matrix, $gdb_list ) = @_;
-	
 	my @gdb_id_list = map {$_->dbID} @$gdb_list;
+	@gdb_id_list = grep { defined && m/[^\s]/ } @gdb_id_list;
 	my $submatrix = Bio::EnsEMBL::Compara::Utils::DistanceMatrix->new();
 	foreach my $gdb1 ( @gdb_id_list ) {
 		foreach my $gdb2 ( @gdb_id_list ) {
+			#print Dumper $submatrix;
+			#print "$gdb1 $gdb2 ".$matrix->distance($gdb1, $gdb2)."\n";
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/Utils/DistanceMatrix.pm]: remove commented lines

File: modules/Bio/EnsEMBL/Compara/Utils/DistanceMatrix.pm [hunk 14]
```diff
@@ -364,4 +368,14 @@ sub phylip_from_matrix {
...
 }
 
+sub empty_submatrix {
+  my ($self, $submatrix) = @_;
+
+  my @members = $submatrix->members;
+  foreach my $member ( @members ) {
+    return 0 unless $member =~ m/^mrg_/;
+  }
+  return 1;
```
Comment (carlacummins) [inline modules/Bio/EnsEMBL/Compara/Utils/DistanceMatrix.pm]: indentation: tabs should be 4 spaces

File: modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm (hunk 11 above)
Comment (CristiGuijarro) [reply modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm]: I will check

Comment (CristiGuijarro) [reply modules/Bio/EnsEMBL/Compara/RunnableDB/SpeciesTree/PermuteMatrix.pm]: It was not needed, I have removed it

File: modules/Bio/EnsEMBL/Compara/PipeConfig/CreateSpeciesTree_conf.pm [hunk 15]
```diff
@@ -67,7 +72,7 @@ sub default_options {
         'master_db'          => 'compara_master',
 
         'representative_species' => undef,
-        'taxonomic_ranks' => ['order', 'class', 'phylum', 'kingdom'],
+        'taxonomic_ranks' => ['genus', 'family', 'order', 'class', 'phylum', 'kingdom'],
```
Comment (muffato) [inline modules/Bio/EnsEMBL/Compara/PipeConfig/CreateSpeciesTree_conf.pm]: Does it work for Vertebrates with both _genus_ and _family_ in the list ?

Comment (CristiGuijarro) [reply modules/Bio/EnsEMBL/Compara/PipeConfig/CreateSpeciesTree_conf.pm]: Yes, a filter is put in later to ensure that for the large species_sets the taxonomic ranks that contain too few species are eliminated. Tests have been run on both the rice species set and the default vertebrates. The default vertebrates tree from these tests and the release 99 tree of default vertebrates have been compared and shown to be identical.

**Pull request merged**